api/desecapi/views.py

```diff
@@ -120,6 +125,90 @@ class DomainDetailByName(DomainDetail):
     lookup_field = 'name'
 
 
+class RRsetDetail(generics.RetrieveUpdateDestroyAPIView):
+    lookup_field = 'type'
+    serializer_class = RRsetSerializer
+    permission_classes = (permissions.IsAuthenticated, IsDomainOwner,)
+    restricted_types = ('SOA', 'RRSIG', 'DNSKEY', 'NSEC3PARAM')
+
+    def delete(self, request, *args, **kwargs):
+        try:
+            super().delete(request, *args, **kwargs)
+        except Http404:
+            pass
+        return Response(status=status.HTTP_204_NO_CONTENT)
+
+    def get_queryset(self):
+        name = self.kwargs['name']
+        subname = self.kwargs['subname'].replace('=2F', '/')
+        type_ = self.kwargs['type']
+
+        # This may be more cleanly implemented using object level permissions.
+        if type_ in self.restricted_types:
+            raise PermissionDenied("You cannot tamper with the %s RRset." % type_)
+
+        return RRset.objects.filter(
+            domain__owner=self.request.user.pk,
+            domain__name=name, subname=subname, type=type_)
+
+    def update(self, request, *args, **kwargs):
+        if request.data.get('records') == []:
+            return self.delete(request, *args, **kwargs)
+
+        try:
+            return super().update(request, *args, **kwargs)
+        except django.core.exceptions.ValidationError as e:
+            ex = ValidationError(detail=e.message_dict)
+            ex.status_code = status.HTTP_409_CONFLICT
+            raise ex
+
+
+class RRsetsDetail(generics.ListCreateAPIView):
+    serializer_class = RRsetSerializer
+    permission_classes = (permissions.IsAuthenticated, IsDomainOwner,)
+
+    def get_queryset(self):
+        rrsets = RRset.objects.filter(domain__owner=self.request.user.pk,
+                                      domain__name=self.kwargs['name'])
+
+        for filter_field in ('subname', 'type'):
+            value = self.request.query_params.get(filter_field)
+
+            if value is not None:
+                # This may be more cleanly implemented using object level permissions.
+                if filter_field == 'type' and value in RRsetDetail.restricted_types:
+                    raise PermissionDenied("You cannot tamper with the %s RRset." % value)
+
+                rrsets = rrsets.filter(**{'%s__exact' % filter_field: value})
+
+        return rrsets
+
+    def create(self, request, *args, **kwargs):
+        if self.kwargs.get('subname') is not None or self.kwargs.get('type') is not None:
+            raise MethodNotAllowed(method=request.method)
+
+        type_ = request.data.get('type', '')
+        if type_ in RRsetDetail.restricted_types:
+            raise PermissionDenied("You cannot tamper with the %s RRset." % type_)
```

Comment:
tamper IMHO does not include just looking at things, i.e. read. "You may not access"?